File: libs/service_layer/search/SearchEngine/src/main/kotlin/net/dankito/service/search/util/CombinedLazyLoadingList.java
package net.dankito.service.search.util;

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;


public class CombinedLazyLoadingList<T> extends AbstractList<T> {

  protected List<Collection<T>> combinedUnderlyingCollections = new ArrayList<>();


  public CombinedLazyLoadingList() {

  }

  public CombinedLazyLoadingList(Collection<T>... underlyingCollections) {
    for(Collection<T> underlyingCollection : underlyingCollections)
      combinedUnderlyingCollections.add(underlyingCollection);
  }


  @Override
  public int size() {
    int size = 0;
    for(Collection<T> underlyingCollection : combinedUnderlyingCollections)
      size += underlyingCollection.size();
    return size;
  }

  @Override
  public T get(int index) {
    int count = 0;
    Collection<T> underlyingCollectionForThisIndex = null;

    for(Collection<T> underlyingCollection : combinedUnderlyingCollections) {
      if(underlyingCollection.size() + count > index) {
        underlyingCollectionForThisIndex = underlyingCollection;
        break;
      }
      count += underlyingCollection.size();
    }

    if(underlyingCollectionForThisIndex != null)
      return getItemFromCollection(index - count, underlyingCollectionForThisIndex);

    return null;
  }

  protected T getItemFromCollection(int index, Collection<T> collection) {
    if(collection instanceof List)
      return ((List<T>) collection).get(index);

    int i = 0;
    Iterator<T> iterator = collection.iterator();
    while(iterator.hasNext()) {
      if(i == index)
        return iterator.next();

      iterator.next();
      i++;
    }

    return null;
  }

  public void setUnderlyingCollection(Collection<T> underlyingCollection) {
    clear();

    combinedUnderlyingCollections.add(underlyingCollection);
  }

  @Override
  public boolean addAll(Collection<? extends T> c) {
    return combinedUnderlyingCollections.add((Collection<T>)c);
  }

  @Override
  public void clear() {
    combinedUnderlyingCollections.clear();
  }
}

File: libs/service_layer/search/SearchEngine/src/main/kotlin/net/dankito/service/search/util/CombinedLazyLoadingList.kt
package net.dankito.service.search.util

import java.util.*


class CombinedLazyLoadingList<T> : AbstractList<T> {

    private var combinedUnderlyingCollections: MutableList<Collection<T>> = ArrayList()


    constructor() {

    }

    constructor(vararg underlyingCollections: Collection<T>) {
        for (underlyingCollection in underlyingCollections)
            combinedUnderlyingCollections.add(underlyingCollection)
    }


    override val size: Int
        get() {
            var size = 0

            combinedUnderlyingCollections.forEach { size += it.size }

            return size
        }

    override fun get(index: Int): T? {
        var count = 0
        var underlyingCollectionForThisIndex: Collection<T>? = null

        for (underlyingCollection in combinedUnderlyingCollections) {
            if(underlyingCollection.size + count > index) {
                underlyingCollectionForThisIndex = underlyingCollection
                break
            }
            count += underlyingCollection.size
        }

        if(underlyingCollectionForThisIndex != null) {
            return getItemFromCollection(index - count, underlyingCollectionForThisIndex)
        }

        return null
    }

    private fun getItemFromCollection(index: Int, collection: Collection<T>): T? {
        if (collection is List<*>) {
            return (collection as List<T>)[index]
        }

        var i = 0
        val iterator = collection.iterator()

        while (iterator.hasNext()) {
            if(i == index) {
                return iterator.next()
            }

            iterator.next()
            i++
        }

        return null
    }

    fun setUnderlyingCollection(underlyingCollection: Collection<T>) {
        clear()

        combinedUnderlyingCollections.add(underlyingCollection)
    }

    override fun addAll(collection: Collection<T>): Boolean {
        return combinedUnderlyingCollections.add(collection)
    }

    override fun clear() {
        combinedUnderlyingCollections.clear()
    }
}
